### PR TITLE
[FRONTEND] Refactor init, categories, search, product card, login, signup, routes, mocks...

### LINKED ISSUE #292: [FRONTEND] Racing condition with /init and /shoppingCart
This is a ***bug*** issue.

## **Description of the bug:**

Sometimes the shopping cart wouldn't be able to load and shows unauthorized error, but a refresh usually fixes it

## **Steps to reproduce the behavior:**

It is hard to describe how to reproduce this because it depends on the timing of the request. If you can make shopping cart request end *before* /init call ends you will see that behavior.

## **Expected behavior:** 

Shopping cart shows cart items properly without an error

## **Actual behavior:** 

Unauthorized error

### LINKED ISSUE #293: [FRONTEND] Cannot search with category or subcategory because of key change
This is a ***bug*** issue.

## **Description of the bug:**

Cannot specify category or subcategory in search side panel due to the change of category/subcategory keys in backend.

## **Steps to reproduce the behavior:**

1. Go to search page
2. Enter a category and/or subcategory
3. Error notification

## **Expected behavior:**

Be able to filter by category

## **Actual behavior:** 

Error from backend

### LINKED ISSUE #294: [FRONTEND] Product card show NaN in place of discounted price
This is a ***bug*** issue.

## **Description of the bug:**

Product card shows NaN for discounted price.

## **Steps to reproduce the behavior:**

1. Go to search page
2. Observe product card

## **Expected behavior:**

Show discounted price correctly

## **Actual behavior:** 

Shows NaN

## Screenshot

image

### LINKED ISSUE #295: [FRONTEND] Missing last name in profile avatar and profile page
This is a ***bug*** issue.

## **Description of the bug:**

Avatar in header and profile page do not show last name

## **Steps to reproduce the behavior:**

1. Login
2. Look at avatar in top-right corner
3. Go to profile page
4. See missing last name

## **Expected behavior:**

Show first name last name

## **Actual behavior:** 

Shows first name only

## Screenshot

image
image

### LINKED ISSUE #296: [FRONTEND] Missing picture link and wrong vendor link in shopping cart item
This is a ***bug*** issue.

## **Description of the bug:**

Clicking on shopping cart item does not bring to product page and clicking on vendor in shopping cart item does not bring to vendor's page

## **Steps to reproduce the behavior:**

1. Login

2. Make sure you have some products in your shopping cart (add them from search for example)

3. Attempt click on picture of shopping cart item

4. Click on vendor link

   

**Expected behavior:**

Picture click leads to product, vendor click brings to vendor page

## **Actual behavior:** 

Picture is not a link and vendor url is wrong therefore does not lead to vendor page

### LINKED ISSUE #297: [FRONTEND] some routes are not protected by auth
This is a ***bug*** issue.

## **Description of the bug:**

Customer users are able to go to signup and login pages and guest users are able to go to shopping cart and checkout pages

## **Steps to reproduce the behavior:**

1. Login
2. Replace url with /signup or /login
3. Logout
4. Replace url with /checkout or /shoppingCart

**Expected behavior:**

Redirect to / on those urls

## **Actual behavior:** 

Actually goes to those page

### LINKED ISSUE #300: [FRONTEND] When price is too long in product card, overflow happens
This is a ***bug*** issue.

## Description of the bug:

Product card becomes poorly formatted when price is too long

## Steps to reproduce the behavior:

1. Go to `/search/products?category=1&pageSize=10&subcategory=2&type=products`
2. Look at last result

## Expected behavior:

Properly formatted price etc

## **Actual behavior: 

Everything breaks into second line

## Screenshot

image

### PR DESCRIPTION
**Pull request by MehdiSaffar**
## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

# Description

### Remove race condition with /init

I made it so that /init and /categories loads first before the app itself uses those values. This solves the shopping cart attempting to load before /init finishes.

### Request interceptor defined in api.js instead

Easiest and safest way to always have the most up to date token value in localStorage.

### Loading screen

While the app is loading, shows the getflix logo over white background with 2s fade-in animations. Loading spinner starts only after 3s

### Categories and subcategories

- Categories are now dynamically loaded from backend during init and stored in app context
- Refactored all components that depend on categories to work with this
- Removed old categories from utils.js
- Search side panel now actually changes subcategories based on currently selected category

### Search page

- Fixed products not loading because of wrong `products` key placement in JSON returned from backend.

### Product card

- Fixed NaN showing on discounted price

### Shopping cart page

- Made product pictures clickable, brings to product page
- Fixed vendor.name being passed to link of vendor instead of vendor.id

### Profile

- Fix last name not showing in avatar
- Fix last name not showing in profile page

### Login/Signup

- Removed mock delay
- Add proper setLoading with try and finally

### Routes

- Protect routes based on user type (guest, customer, vendor)
- Add proper redirect when wrong url typed

### Mocks

- Add mock for /init
- Add mock for /categories
- Add warning on development environment showing how many active mock handlers are active (to prevent mistakes)
- All mocks are deactivated in this PR

# Screenshots

### Loading screen
image

### Last name showing properly
image
image

### Product card NaN fix
image

# Test results

Please run all the tests and dump the result/summary here in a code block.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have checked my code and corrected any misspellings
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

# Potential problems/Important remarks

- Added warning comment in profile page about the name of the key being surname instead of lastname

fixes #297 
fixes #296 
fixes #295
fixes #294
fixes #293

### CODE REVIEW

File: ecommerce/frontend/src/components/product_card/ProductCard.js [hunk 1]
```diff
@@ -15,9 +15,10 @@ export const ProductCard = ({ product, width = 350 }) => {
 
     const onAddToList = product => {}
 
-    const { title, rating, price, currency, imageUrl, id, discount } = product
+    const { title, rating, price, price_after_discount, currency, imageUrl, id, discount } = product
 
-    const discountPrice = price * (1 - discount)
+    // const discountPrice = price * (1 - discount)
+    console.log(product)
```
Comment (abdullahyldz) [inline ecommerce/frontend/src/components/product_card/ProductCard.js]: In the Cell Phones and Accessories subcategory, there returns 2 items like this 
image
Their passed fields in the ProductCard is like this:
```
Object
currency: "TL"
id: 22
imageUrl: undefined
price: 799.9
price_after_discount: 703.912
rating: 0
title: "Los Angeles Lakers Home"
__proto__: Object
```

```
Object
currency: "TL"
id: 26
imageUrl: "/image/3"
price: 799.9
price_after_discount: 703.912
rating: 0
title: "Los Angeles Lakers Home Jersey"
__proto__: Object
```

I think,  second imageUrl is the `new format `Backend sends us such that we append it after Backend URL. Other image URLs are just correct. Can you make sure?

Comment (abdullahyldz) [reply ecommerce/frontend/src/components/product_card/ProductCard.js]: Also, another problem is same item is returned again with the same query.

image

File: ecommerce/frontend/src/components/pages/SearchPage.js [hunk 2]
```diff
@@ -8,7 +8,7 @@ import { Helmet } from 'react-helmet'
 import { useHistory } from 'react-router-dom'
 
 import { api } from '../../api'
-import { categories, subcategories } from '../../utils'
+// import { categories, subcategories } from '../../utils'
```
Comment (abdullahyldz) [inline ecommerce/frontend/src/components/pages/SearchPage.js:11]: [SOLVED] Clicking the main category does not properly set URL with category ID. 
Solved it in commit : ad4bda3

image

Comment (abdullahyldz) [reply ecommerce/frontend/src/components/pages/SearchPage.js:11]: Thus, breaks the Backend with 500 (Internal Error)

File: ecommerce/frontend/src/components/layout/Content.js [hunk 3]
```diff
@@ -13,21 +13,28 @@ import { ShoppingCartPage } from '../ShoppingCart/ShoppingCartPage'
...
                 <Route path="/search/:type" component={SearchPage} />
                 <Route path="/product/:productId" component={ProductPage} />
-                <Route path="/" component={HomePage} />
+                {isGuest && <Route path="/login" component={LoginPage} />}
+                {isCustomer && <Route path="/checkout" component={CheckoutPage} />}
+                {isCustomer && <Route path="/shoppingCart" component={ShoppingCartPage} />}
+                {isGuest && <Route path="/signup" component={SignupPage} />}
+                <Route path="/profile" component={ProfilePage} />
+                <Route exact path="/" component={HomePage} />
+                <Redirect to="/" />
```
Comment (abdullahyldz) [inline ecommerce/frontend/src/components/layout/Content.js:37]: These were needed, thanks.

**Review by abdullahyldz — CHANGES_REQUESTED** [review]
I reported some bugs and corrected two small bugs.
Very useful improvements to project.

File: ecommerce/frontend/src/components/product_card/ProductCard.js (hunk 1 above)
Comment (MehdiSaffar) [reply ecommerce/frontend/src/components/product_card/ProductCard.js]: @abdullahyldz two of products returned from backend have an empty images array, and one with `/image/3`. 

image


Product should never be allowed not to have pictures.

For `/image/3` I guess we need to prefix ourselves. I will see what I can do about that

**Comment by MehdiSaffar** [comment]
fixes #300 

image

also fixed image not showing by prefixing with apiUrl

**Comment by MehdiSaffar** [comment]
Integrated all changes, merging

**Pull request merged**